hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/WritableUpgradeStore.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (C) 2023 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.node.app.service.file.impl;
+
+import static java.util.Objects.requireNonNull;
+
+import com.hedera.hapi.node.state.file.File;
+import com.hedera.node.app.spi.state.WritableQueueState;
+import com.hedera.node.app.spi.state.WritableSingletonState;
+import com.hedera.node.app.spi.state.WritableStates;
+import com.hedera.pbj.runtime.io.buffer.Bytes;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.util.function.Predicate;
+
+/**
+ * Provides write methods for modifying underlying data storage mechanisms for working with Files.
+ *
+ * <p>This class is not exported from the module. It is an internal implementation detail. This
+ * class is not complete, it will be extended with other methods like remove, update etc.,
+ */
+public class WritableUpgradeStore extends ReadableUpgradeStoreImpl implements WritableQueueState<File> {
+    /** The underlying data storage class that holds the file data. */
+    private final WritableQueueState<Bytes> writableUpgradeState;
+
+    private final WritableSingletonState<File> writableUpgradeFileState;
+
+    private static final Predicate<Bytes> TRUE_PREDICATE = new TruePredicate();
+
+    /**
+     * Create a new {@link WritableUpgradeStore} instance.
+     *
+     * @param states The state to use.
+     */
+    public WritableUpgradeStore(@NonNull final WritableStates states) {
+        super(states);
+        this.writableUpgradeState = requireNonNull(states.getQueue(getStateKey()));
+        this.writableUpgradeFileState = requireNonNull(states.getSingleton(getFileStateKey()));
+    }
+
+    public void add(@NonNull File file) {
+        requireNonNull(file);
+        writableUpgradeState.add(file.contents());
+        writableUpgradeFileState.put(file);
+    }
+
+    public void resetFileContents() {
+        while (writableUpgradeState.removeIf(TRUE_PREDICATE) != null)
```

Comment:
This doesn't need to be inside a `while` loop any more, right?

Reply:
I'm not sure if I get it right, but we clear the queue before we add a new file during file upgrade

Reply:
Oh n/m the semantics of this `removeIf()` are different than the `j.u.Collection.removeIf()`. 😝